Kubelet should honour the VolumeAttributes which are reported by the volume plugin

carlory opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

**pods.yaml**

```
apiVersion: v1
kind: Pod
metadata:
  name: pod1
  namespace: default
spec:
  containers:
  - image: registry.k8s.io/e2e-test-images/echoserver:2.3
    imagePullPolicy: IfNotPresent
    name: test
    volumeMounts:
    - mountPath: /volume
      name: volume
  volumes:
  - image:
      pullPolicy: IfNotPresent
      reference: quay.io/crio/artifact:v1
    name: volume
---
apiVersion: v1
kind: Pod
metadata:
  name: pod2
  namespace: default
spec:
  containers:
  - image: registry.k8s.io/e2e-test-images/echoserver:2.3
    imagePullPolicy: IfNotPresent
    name: test
    volumeMounts:
    - mountPath: /volume
      name: volume
  volumes:
  - image:
      pullPolicy: IfNotPresent
      reference: quay.io/crio/artifact:v1
    name: volume
---
apiVersion: v1
kind: Pod
metadata:
  name: pod3
  namespace: default
spec:
  containers:
  - image: registry.k8s.io/e2e-test-images/echoserver:2.3
    imagePullPolicy: IfNotPresent
    name: test
    volumeMounts:
    - mountPath: /etc/hosts
      name: volume
  volumes:
  - image:
      pullPolicy: IfNotPresent
      reference: quay.io/crio/artifact:v1
    name: volume
---
apiVersion: v1
kind: Pod
metadata:
  name: pod4
  namespace: default
spec:
  containers:
  - image: registry.k8s.io/e2e-test-images/echoserver:2.3
    imagePullPolicy: IfNotPresent
    name: test
    volumeMounts:
    - mountPath: /etc/hosts
      name: volume
  volumes:
  - emptyDir: {}
    name: volume
```

- pod1 and pod2 share the same `image` volume source type and are mounted to `/volume` in the container.
- pod3 and pod4 mount the volume to `/etc/hosts` in the container but use different volume source types.

**without this fix** 

pod3 is running but pod4 has `CreateContainerError` status.

```
(base) ➜  kubernetes git:(fix-image-volume-mount) kubectl get po
NAME   READY   STATUS                 RESTARTS   AGE
pod1   1/1     Running                0          29s
pod2   1/1     Running                0          29s
pod3   1/1     Running                0          29s
pod4   0/1     CreateContainerError   0          29s
```

pod1 mount with `rw` option

```
root@crio2-control-plane:/# crictl inspect 5c76497deb40c |  jq .status.mounts[0]
{
  "containerPath": "/volume",
  "gidMappings": [],
  "hostPath": "/var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged",
  "propagation": "PROPAGATION_PRIVATE",
  "readonly": false,
  "recursiveReadOnly": false,
  "selinuxRelabel": false,
  "uidMappings": []
}

root@crio2-control-plane:/# crictl inspect 5c76497deb40c | jq .info.runtimeSpec.mounts[10]
{
  "destination": "/volume",
  "type": "bind",
  "source": "/var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged",
  "options": [
    "rbind",
    "rprivate",
    "rw",
    "bind"
  ]
}
```

**with this fix** 

pod3 and pod4 has `CreateContainerError` status.

```
(base) ➜  kubernetes git:(fix-image-volume-mount) kubectl get po
NAME   READY   STATUS                 RESTARTS   AGE
pod1   1/1     Running                0          44m
pod2   1/1     Running                0          44m
pod3   0/1     CreateContainerError   0          44m
pod4   0/1     CreateContainerError   0          44m

(base) ➜  kubernetes git:(fix-image-volume-mount) kubectl describe po pod3
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  51m                  default-scheduler  Successfully assigned default/pod3 to crio-control-plane
  Normal   Pulled     50m (x8 over 51m)    kubelet            Container image "registry.k8s.io/e2e-test-images/echoserver:2.3" already present on machine
  Warning  Failed     50m (x8 over 51m)    kubelet            Error: container create failed: mount `/var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged` to `etc/hosts`: Not a directory
  Normal   Pulled     78s (x229 over 51m)  kubelet            Container image "quay.io/crio/artifact:v1" already present on machine

(base) ➜  kubernetes git:(fix-image-volume-mount) kubectl describe po pod4
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  51m                  default-scheduler  Successfully assigned default/pod4 to crio-control-plane
  Warning  Failed     49m (x12 over 51m)   kubelet            Error: container create failed: mount `/var/lib/kubelet/pods/983847e3-109e-4a2e-89f6-5cd0fef3b4cc/volumes/kubernetes.io~empty-dir/volume` to `etc/hosts`: Not a directory
  Normal   Pulled     81s (x233 over 51m)  kubelet            Container image "registry.k8s.io/e2e-test-images/echoserver:2.3" already present on machine
```

pod1 mount with `ro` option

```
(base) ➜  kubernetes git:(fix-image-volume-mount) docker exec -it 92a9726219db bash
root@crio-control-plane:/# crictl inspect 90e489c2aa5e2 | jq .status.mounts[0]
{
  "containerPath": "/volume",
  "gidMappings": [],
  "hostPath": "/var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged",
  "propagation": "PROPAGATION_PRIVATE",
  "readonly": true,
  "recursiveReadOnly": false,
  "selinuxRelabel": false,
  "uidMappings": []
}

root@crio-control-plane:/# crictl inspect 90e489c2aa5e2 | jq .info.runtimeSpec.mounts[10]
{
  "destination": "/volume",
  "type": "bind",
  "source": "/var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged",
  "options": [
    "rbind",
    "rprivate",
    "ro",
    "bind"
  ]
}
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related to https://github.com/kubernetes/enhancements/issues/4639

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
1. When the kubelet constructs the cri mounts for the container which references an `image` volume source type, It passes the missing mount attributes to the CRI implementation, including `readOnly`, `propagation`, and `recursiveReadOnly`. When the readOnly field of the containerMount is explicitly set to false, the kubelet will take the `readOnly`as true to the CRI implementation because the image volume plugin requires the mount to be read-only. 
2. Fix a bug where the pod is unexpectedly running when the `image` volume source type is used and mounted to `/etc/hosts` in the container.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

carlory commented:
/hold

carlory commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -291,69 +279,83 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			vol.SELinuxLabeled = true
 			relabelVolume = true
```
Do we want to relabel the image volume when the SELinux is enabled on the node?

```go
func (o *imagePlugin) GetAttributes() volume.Attributes {
	return volume.Attributes{
		ReadOnly:       true,
		Managed:        true,
		SELinuxRelabel: true,
	}
}
```

BTW, the image volume is shared among all the pods on the node if the pod's volume is an image. I'm unfamiliar with SELinux, what's the impact if we relabel the image volume?

When SELinux is enabled, can pod1 and pod2 access the same image volume successfully even if Managed is false?

carlory replied:
I can not enable selinux on kind node. I'm not sure whether the kind does not support selinux.

What's the expected behavior on image volume when SELinux is enabled on node? @saschagrunert

saschagrunert replied:
Let's pass that information to the runtime and let them decide. For CRI-O we mount directly using the correct mount label, so the indicator should not change anything. 

> When SELinux is enabled, can pod1 and pod2 access the same image volume successfully even if Managed is false

Not when they use different labels, which is a known limitation for volume plugins.

carlory replied:
Does the pod1 and pod2 have the same mount label (selinux)? 

I see that the image volume has already be labeled when the first pod is created.

https://github.com/cri-o/cri-o/blob/485358b6ef0b69b0a345b3d4e5c62211477aeb8a/server/container_create_linux.go#L1078

saschagrunert replied:
> Does the pod1 and pod2 have the same mount label (selinux)?

Depends on the platform, but does that matter? We have to fail pod2 if pod1 is already using a different label.

carlory replied:
> Not when they use different labels, which is a known limitation for volume plugins.

Does the kubernetes website include the known limitation? If not, it is better to add it. Can you update the website to include the known limitation? I'm worried that I can't explain it clearly.

saschagrunert replied:
Yes, I can take care of that.

saschagrunert replied:
Closing the loop: https://github.com/kubernetes/website/pull/47588

carlory replied:
> Yes, I can take care of that.

Thank you.

> Let's pass that information to the runtime and let them decide.

Do we need to set Managed and SELinuxRelabel to false in `GetAttributes` of the image plugin?

the hostPath plugin set these values to false

```go
func (b *hostPathMounter) GetAttributes() volume.Attributes {
	return volume.Attributes{
		ReadOnly:       b.readOnly,
		Managed:        false,
		SELinuxRelabel: false,
	}
}
```

saschagrunert replied:
I'm not sure honestly, my instinct would say set them to `true` (as-is).

carlory replied:
> Closing the loop: https://github.com/kubernetes/website/pull/47588

/cc @kerthcet 

When scheduling pods, does the scheduler need to add a new plug-in to adapt to this situation? To avoid the pod from not being able to start up due to SELinux being enabled after the pod is scheduled to the node

/cc @wzshiming 

Does the containerd has the same limitation like cri-o?

carlory replied:
> I'm not sure honestly, my instinct would say set them to true (as-is).

cc @jsafrane @gnufied can you have a look?

carlory replied:
@saschagrunert It seems that we can not relabel the volume when the selinux is enabled.

```
              waiting:
                message: 'relabel failed /var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged:
                  lsetxattr /var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged/dir:
                  read-only file system'
                reason: CreateContainerError
```

saschagrunert replied:
We probably have to considering the test failures in `pull-kubernetes-node-crio-cgrpv2-imagevolume-`e2e

carlory replied:
Is it possible for image volume to make it act like emptyDir in the cri-o? 

One pod/containers can isolate the volume from another pod/containers. Don't share the volume between different pods/containers. They can have different mountpoints for the same image.

carlory replied:
If we can do that, no other requirement will be needed in kube-scheduler. Users can use it more easily without extra configuration.

saschagrunert replied:
@carlory we need to extend the c/storage API to be able to specify a different mountpoint.

saschagrunert replied:
But I'm not sure if we can even run into that case, see: https://github.com/containers/storage/issues/1994

carlory replied:
How about this? @saschagrunert  

1. Extend the `kuberuntime.ImageService` proto to support mount/unmount image: (I'm not sure whether it can be done)

```
// ImageService defines the public APIs for managing images.
service ImageService {
    // MountImage mount the image to the given path.
    rpc MountImage(MountImageRequest) returns (MountImageResponse) {}
    // UnmountImage unmount the image from the given path.
    rpc UnmountImage(UnmountImageRequest) returns (UnmountImageResponse) {}
}
```

2. Image volume plugin:

```go
type imageVolume struct {
	volName string
	podUID  types.UID
	plugin  *imagePlugin
	volume.MetricsProvider
}

var _ volume.Volume = &imageVolume{} 

// allow volume plugin to manage the volume path.
func (iv *imageVolume) GetPath() string { 
	return iv.plugin.host.GetPodVolumeDir(iv.podUID, utilstrings.EscapeQualifiedName(pluginName), iv.volName)
}

type imageVolumeMounter struct {
	*imageVolume

	source v1.ImageVolumeSource
	pod    v1.Pod

	pullImage    func(pod v1.Pod, source v1.ImageVolumeSource) error
	mountImage   func(path string, image string) error // call cri to mount image
	unmountImage func(path string) error // call cri to unmount image
}

var _ volume.Mounter = &imageVolumeMounter{}

func (m *imageVolumeMounter) RequiresFSResize() bool { return false }

func (m *imageVolumeMounter) RequiresRemount(spec *volume.Spec) bool { return false }

func (m *imageVolumeMounter) GetAttributes() volume.Attributes {
	return volume.Attributes{
		ReadOnly:       true,
		Managed:        true,
		SELinuxRelabel: true,
	}
}

func (m *imageVolumeMounter) SetUp(mounterArgs volume.MounterArgs) error {
	return m.SetUpAt(m.GetPath(), mounterArgs)
}

func (m *imageVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterArgs) error {
	if err := m.pullImage(m.pod, m.source); err != nil {
		return err
	}
	return m.mountImage(dir, m.source.Reference)
}

func (m *imageVolumeMounter) SupportsMountOption() bool { return false }

func (m *imageVolumeMounter) SupportsSELinuxContextMount(spec *volume.Spec) (bool, error) {
	return false, nil
}

func (m *imageVolumeMounter) TearDown() error {
	return m.TearDownAt(m.GetPath())
}

func (m *imageVolumeMounter) TearDownAt(string) error {
	return m.unmountImage(m.GetPath())
}
```
If we can do, 

- maybe we able to support subpath in the future?
- allow more than one pods to running on the same node when using the same image.
- remove specific logic in kubelet pod loop and `makemounts`.

saschagrunert replied:
@carlory I think changes like this needs to go into a graduation of the feature and are out of scope of this fix here. I appreciate your input here :pray:

carlory replied:
It's indeed out of scope of this fix. I will mark the above conversation as resolved. 

And if there are discussions on this topic elsewhere in the future, I would be happy to participate if possible.  😊

carlory commented:
/cc @jsafrane @gnufied @saschagrunert

saschagrunert commented:
> The refCount of the layer is being rapidly increased. What's wrong with my changes?

Do the CRI-O `--log-level=debug` logs indicate multiple calls to `Store().MountImage(…)` by:

```
Image ref to mount: …
```
and/or
```
Image ID to mount: …
```

saschagrunert commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -291,69 +279,83 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+				err = volumevalidation.ValidatePathNoBacksteps(subPath)
 				if err != nil {
-					return nil, cleanupAction, err
+					return nil, cleanupAction, fmt.Errorf("unable to provision SubPath `%s`: %v", subPath, err)
```
nit, wanna wrap here?
```suggestion
					return nil, cleanupAction, fmt.Errorf("unable to provision SubPath `%s`: %w", subPath, err)
```

carlory commented:
crio log: 

```
root@crio-control-plane:/run/containers/storage/overlay-layers# cat mountpoints.json | jq .[0]
{
  "id": "82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618",
  "path": "/var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged",
  "count": 620
}
root@crio-control-plane:/run/containers/storage/overlay-layers# journalctl -u crio -e | grep 'Image ref to mount' | wc -l
5
root@crio-control-plane:/run/containers/storage/overlay-layers# journalctl -u crio -e | grep 'Image ID to mount' | wc -l
5
```

saschagrunert commented:
/test

saschagrunert commented:
/test pull-kubernetes-node-crio-cgrpv2-imagevolume-e2e

carlory commented:
> we may have to change something SELinux related:

okay, I disabled the relabel and manage field

carlory commented:
/test pull-kubernetes-node-crio-cgrpv2-imagevolume-e2e

saschagrunert commented:
/test pull-kubernetes-node-crio-cgrpv2-imagevolume-e2e

gnufied commented:
Just to clarify - this appears to be only a problem if we are mounting `/etc/hosts` right? Other mount points work fine.

carlory commented:
> Just to clarify - this appears to be only a problem if we are mounting `/etc/hosts` right? Other mount points work fine.澄清一下 - 这似乎只是在我们安装 `/etc/hosts` 时才有问题，对吧？其他挂载点工作正常。

Yes, it is

carlory commented:
/hold cancel

gnufied commented:
This looks good from storage side. But it is not entirely clear that what should be expected behaviour when this happens, assuming no selinux context related errors are involved. cc @saschagrunert

carlory commented:
run again because the job is updated.
/test pull-kubernetes-node-crio-cgrpv2-imagevolume-e2e

carlory commented:
PTAL @kubernetes/sig-node-pr-reviews

carlory commented on pkg/volume/image/image.go:
```diff
@@ -67,8 +67,8 @@ func (o *imagePlugin) ConstructVolumeSpec(volumeName, mountPath string) (volume.
 func (o *imagePlugin) GetAttributes() volume.Attributes {
 	return volume.Attributes{
 		ReadOnly:       true,
-		Managed:        true,
```
Please double-check for SELinux management, do we need to reset them to `true`? @saschagrunert

saschagrunert replied:
I think `Managed` only enables this code path: 

https://github.com/kubernetes/kubernetes/blob/92507eaabe5c93158f64faca240d02d0e5875a01/pkg/kubelet/kubelet_pods.go#L290-L293

This makes it possible to set the `SelinuxRelabel` field for the runtime in:

https://github.com/kubernetes/kubernetes/blob/92507eaabe5c93158f64faca240d02d0e5875a01/pkg/kubelet/kuberuntime/kuberuntime_container.go#L442-L451

In CRI-O we don't need to relabel because we can mount with the right label from the process. Means that setting it to `false` seems reasonable.

I assume the e2e tests fail if we set this field to `true`?

carlory commented:
/cc @SergeyKanzhelev

SergeyKanzhelev commented:
for release notes, I would suggest to rephrase the first one from the `SHOULD` statement to the description on what was changed

SergeyKanzhelev commented:
How any of those changes tested? Why no existing tests broke? Can you please add tests demonstrating the wrong behavior and the fix

SergeyKanzhelev commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -291,69 +279,83 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			vol.SELinuxLabeled = true
 			relabelVolume = true
 		}
-		hostPath, err := volumeutil.GetPath(vol.Mounter)
```
I am concerned of moving and changing so much code that is not under the feature gate. Was it added alongside the feature?

Ideally we want to minimize the number of potential regressions and keep existing codepaths as close to what they were without the feature gate as possible

carlory replied:
will update

carlory replied:
> I am concerned of moving and changing so much code that is not under the feature gate. Was it added alongside the feature?

It just adds a new conditon check for the previous code. It doesn't change the code itself. I add the feature-gate check in line 289. If the feature-gate is disabled, the image will be nil and old logic will be executed.  @SergeyKanzhelev

SergeyKanzhelev commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -381,6 +383,7 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			Name:              mount.Name,
 			ContainerPath:     containerPath,
 			HostPath:          hostPath,
+			Image:             image,
```
I wonder how it worked without the image before? What is the image field is being used for if it was not passed here before and nothing broke?

carlory replied:
It aims to add all missing mount information to the runtime and let them decide when the volume source is image.

carlory replied:
@SergeyKanzhelev  some volume attributes reported by the `image` volume plugin, i.e. relabel, read-only, etc., aren't sent to the implementation of CRI runtime in [previous kubelet implementation](https://github.com/kubernetes/kubernetes/pull/125663).  

```go
func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, hostDomain string, podIPs []string, podVolumes kubecontainer.VolumeMap, hu hostutil.HostUtils, subpather subpath.Interface, expandEnvs []kubecontainer.EnvVar, supportsRRO bool, imageVolumes kubecontainer.ImageVolumes) ([]kubecontainer.Mount, func(), error) {
	mountEtcHostsFile := shouldMountHostsFile(pod, podIPs)
	klog.V(3).InfoS("Creating hosts mount for container", "pod", klog.KObj(pod), "containerName", container.Name, "podIPs", podIPs, "path", mountEtcHostsFile)
	mounts := []kubecontainer.Mount{}
	var cleanupAction func()
	for i, mount := range container.VolumeMounts {
		// Check if the mount is referencing an OCI volume
		if imageVolumes != nil && utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) {
			if image, ok := imageVolumes[mount.Name]; ok {
				mounts = append(mounts, kubecontainer.Mount{
					Name:          mount.Name,
					ContainerPath: mount.MountPath,
					Image:         image,
				})
				continue
			}
		}
                 ...skipped..
		mounts = append(mounts, kubecontainer.Mount{
			Name:              mount.Name,
			ContainerPath:     containerPath,
			HostPath:          hostPath,
			ReadOnly:          mount.ReadOnly || mustMountRO,
			RecursiveReadOnly: rro,
			SELinuxRelabel:    relabelVolume,
			Propagation:       propagation,
		})
	}
                 ...skipped..
	return mounts, cleanupAction, nil
```

What's the result?  Please see the `without this fix` section in my PR's description.

gnufied commented:
After SELinux stuff was fixed in cri-o, what is the new behaviour? Is this still the same?

carlory commented:
> I wonder how it worked without the image before? What is the image field is being used for if it was not passed here before and nothing broke?  

@SergeyKanzhelev  some volume attributes reported by the `image` volume plugin, i.e. relabel, read-only, etc., aren't sent to the implementation of CRI runtime in [previous kubelet implementation (https://github.com/kubernetes/kubernetes/pull/125663).  

```go
func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, hostDomain string, podIPs []string, podVolumes kubecontainer.VolumeMap, hu hostutil.HostUtils, subpather subpath.Interface, expandEnvs []kubecontainer.EnvVar, supportsRRO bool, imageVolumes kubecontainer.ImageVolumes) ([]kubecontainer.Mount, func(), error) {
	mountEtcHostsFile := shouldMountHostsFile(pod, podIPs)
	klog.V(3).InfoS("Creating hosts mount for container", "pod", klog.KObj(pod), "containerName", container.Name, "podIPs", podIPs, "path", mountEtcHostsFile)
	mounts := []kubecontainer.Mount{}
	var cleanupAction func()
	for i, mount := range container.VolumeMounts {
		// Check if the mount is referencing an OCI volume
		if imageVolumes != nil && utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) {
			if image, ok := imageVolumes[mount.Name]; ok {
				mounts = append(mounts, kubecontainer.Mount{
					Name:          mount.Name,
					ContainerPath: mount.MountPath,
					Image:         image,
				})
				continue
			}
		}
                 ...skipped..
		mounts = append(mounts, kubecontainer.Mount{
			Name:              mount.Name,
			ContainerPath:     containerPath,
			HostPath:          hostPath,
			ReadOnly:          mount.ReadOnly || mustMountRO,
			RecursiveReadOnly: rro,
			SELinuxRelabel:    relabelVolume,
			Propagation:       propagation,
		})
	}
                 ...skipped..
	return mounts, cleanupAction, nil
```

What's the result?  Please see the `without this fix` section in my PR's description.

> After bind-mount stuff was fixed for selinux stuff - https://github.com/cri-o/cri-o/pull/8521, what is the new behaviour? @carlory can you please confirm without your change and with your change?

@gnufied
This PR is not related to selinux issue, but that is detected by this PR when running the ` pull-kubernetes-node-crio-cgrpv2-imagevolume-e2e` job, in order to make the job pass, the PR changes the volume attributes reported by the `image` volume plugin, see https://github.com/kubernetes/kubernetes/pull/126806#discussion_r1723082367 for more details. At that time, we think the issue is a known limitation for all volume plugins, so we open a PR for the website to document it. We also asked about the limitation with the sig-storage in the community meeting but got a different answer. Thanks a lot for all the help from the community. @saschagrunert raised a PR to fix it in the CRI-O repo and https://github.com/kubernetes/website/pull/47588 was closed, so no new behavior is introduced and https://github.com/kubernetes/kubernetes/pull/126991 added a new test case for node e2e tests to verify the fix. The result is passed. 

since https://github.com/cri-o/cri-o/pull/8521 is merged, the root cause is fixed, so it won't occur again with or without this PR.  Why not revert selinux configuration in this PR, please see https://github.com/kubernetes/kubernetes/pull/126806#discussion_r1749483999. 

This PR mainly aims to add all missing mount information to the runtime and let them decide.

carlory commented:
> How any of those changes tested? Why no existing tests broke? Can you please add tests demonstrating the wrong behavior and the fix

No existing test is broken. I am adding new test.

carlory commented:
> for release notes, I would suggest to rephrase the first one from the `SHOULD` statement to the description on what was changed

updated.

carlory committed:
TestMakeMounts: add new cases for the image volume feature

SergeyKanzhelev commented:
lost track of this PR. Now it only have test files in it. Perhaps a wrong force push?

carlory committed:
1. When the kubelet constructs the cri mounts for the container which references an `image` volume source type, It passes the missing mount attributes to the CRI implementation, including `readOnly`, `propagation`, and `recursiveReadOnly`. When the readOnly field of the containerMount is explicitly set to false, the kubelet will take the `readOnly`as true to the CRI implementation because the image volume plugin requires the mount to be read-only.
2. Fix a bug where the pod is unexpectedly running when the `image` volume source type is used and mounted to `/etc/hosts` in the container.

carlory commented:
/cc @saschagrunert @SergeyKanzhelev @gnufied 

Please review it again. I update the code to simplify the code and make it more readable.

carlory commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -320,7 +308,7 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 		}
 
-		if subPath != "" {
+		if subPath != "" && hostPath != "" {
```
If we want to support the subpath feature for image volume, we can change it like this in the kubelet after it is merged:

```go
var image *runtimeapi.ImageSpec
var imageSubPath string
if imageVolumes != nil && utilfeature.DefaultFeatureGate.Enabled(features.ImageVolume) {
    image = imageVolumes[mount.Name]
    imageSubPath = subPath
}
```

carlory replied:
> if subPath != "" && hostPath != "" {

We don't need to check the volume type because all traditional volume plugins always return a non-empty host path.

carlory committed:
fix pull-kubernetes-linter-hints

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

thank you for resolving all comments
